staging/src/k8s.io/component-base/logs/logrotation/logrotation.go

```diff
@@ -0,0 +1,222 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logrotation
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+)
+
+const timeLayout string = "20060102-150405"
+
+type RotationFile struct {
+	// required, the max size of the log file in MB, 0 means no rotation
+	maxSize uint
+	// required, the max age of the log file in days, 0 means no cleanup
+	maxAge      uint
+	filePath    string
+	mut         sync.Mutex
+	file        *os.File
+	currentSize int64
+	isFlushed   bool
+	stop        chan struct{}
+}
+
+func Open(filePath string, maxSize uint, maxAge uint) (w *RotationFile, err error) {
+	w = &RotationFile{
+		filePath: filePath,
+		maxSize:  maxSize,
+		maxAge:   maxAge,
+	}
+
+	logFile, err := os.OpenFile(w.filePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return nil, err
+	}
+
+	w.file = logFile
+
+	w.isFlushed = true
+	interval := time.NewTicker(time.Second * 5)
+	w.stop = make(chan struct{})
+
+	go func() {
+		defer interval.Stop()
+		for {
+			select {
+			case <-interval.C:
+				// This block runs every time the ticker ticks
+				w.isFlushed = false
+			case <-w.stop:
+				// Stop signal received, exiting the goroutine
+				return
+			}
+		}
+	}()
+
+	if w.maxSize > 0 {
+		info, err := os.Stat(w.filePath)
+		if err != nil {
+			return nil, err
+		}
+		w.currentSize = info.Size()
+	}
+
+	return w, nil
+}
+
+// write func to satisfy io.writer interface
+func (w *RotationFile) Write(p []byte) (n int, err error) {
+	w.mut.Lock()
+	defer w.mut.Unlock()
+
+	n, err = w.file.Write(p)
+	if err != nil {
+		return 0, err
+	}
+
+	if !w.isFlushed {
+		err = w.file.Sync()
+		if err != nil {
+			return 0, err
+		}
+		w.isFlushed = true
+	}
+
+	if w.maxSize > 0 {
+		w.currentSize += int64(len(p))
+
+		// if file size over maxsize rotate the log file
+		if (w.currentSize / 1000000) >= int64(w.maxSize) {
```

Comment:
The problem is that the file size check for rotation is incorrect because it divides the size by 1,000,000 instead of using the correct conversion for megabytes (1 MB = 1024*1024 bytes).  In addition, using division for byte conversion can result in an integer division error; instead, multiplication can be used here.
		if w.currentSize >= int64(w.maxSize)*1024*1024 {
```